### PR TITLE
Add a limits config option to allow for dropping of the cluster label

### LINKED ISSUE #1724: Add ability to drop cluster label after dedupe
A couple of users have asked for this; as we need a cluster label for deduping in the distributors, some people don't have this and adding it will break their graphs.

By "break their graphs", they will cause things like discontinuations in things like recording rules, which for LTS systems is undesirable.

@cstyan is this something we can make an option?

### PR DESCRIPTION
**Pull request by cstyan**
Fixes #1724 

Adds an optional flag per user to drop the cluster label, in addition to dropping the replica label, when ingesting HA samples. Should only be used if that user has their own set of more than two labels that they use to uniquely identify Prometheus replicas.

There's a bit of a refactor here as well:
- I noticed that in the config the ordering of the values was cluster label then replica label, but in RegisterFlags the order is replica then cluster. I moved them around in RegisterFlags to match the order in the struct.
- `distributor.Push` was getting pretty long, I moved the internals of the validation loop into a function `validateSeries`. This also had the side effect of making that portion of the code testable. I've only added tests for the label removal bits. I assume the actual validation functions `ValidateLabels` and `ValidateSamples` are tested already.
- renamed removeReplicaLabel -> removeLabel, it wasn't doing anything replica label specific anyways

Signed-off-by: Callum Styan <callumstyan@gmail.com>
cc @tomwilkie @gouthamve

### CODE REVIEW

File: docs/arguments.md [hunk 1]
```diff
@@ -144,11 +144,13 @@ prefix these flags with `distributor.ha-tracker.`
 
 ### HA Tracker
 
-HA tracking has two of it's own flags:
+HA tracking has three of it's own flags:
```
Comment (bboreham) [inline docs/arguments.md]:
```suggestion
HA tracking has three of its own flags:
```

**Comment by bboreham** [comment]
I find it easier to review PRs like this when you split the pure-refactoring part from the behaviour-changing part, into separate commits. Otherwise I have to figure out your intention on every line and that takes longer.

File: pkg/distributor/distributor.go [hunk 2]
```diff
@@ -280,14 +281,55 @@ func (d *Distributor) checkSample(ctx context.Context, userID, cluster, replica
 	return true, nil
 }
 
+// Validates a single series from a write request. Will remove HA labels if necessary.
```
Comment (bboreham) [inline pkg/distributor/distributor.go]: State what the return values are.
Include why we have return values both via pointer and by value.

File: pkg/util/validation/limits.go [hunk 3]
```diff
@@ -48,8 +49,9 @@ func (l *Limits) RegisterFlags(f *flag.FlagSet) {
...
 	f.BoolVar(&l.AcceptHASamples, "distributor.ha-tracker.enable-for-all-users", false, "Flag to enable, for all users, handling of samples with external labels identifying replicas in an HA Prometheus setup.")
-	f.StringVar(&l.HAReplicaLabel, "distributor.ha-tracker.replica", "__replica__", "Prometheus label to look for in samples to identify a Prometheus HA replica.")
 	f.StringVar(&l.HAClusterLabel, "distributor.ha-tracker.cluster", "cluster", "Prometheus label to look for in samples to identify a Prometheus HA cluster.")
+	f.StringVar(&l.HAReplicaLabel, "distributor.ha-tracker.replica", "__replica__", "Prometheus label to look for in samples to identify a Prometheus HA replica.")
+	f.BoolVar(&l.HADropClusterLabel, "distributor.ha-tracker.drop-cluster-label", false, "Enable this flag to drop the cluster label when ingesting HA samples for a user, in addition to dropping the replica label. Note that this flag should only be used when users already have their own combination of more than two labels to uniquely identify Prometheus replicas.")
```
Comment (bboreham) [inline pkg/util/validation/limits.go]: The 'when' part doesn't appear to be true - it is done unconditionally.

Comment (cstyan) [reply pkg/util/validation/limits.go]: fixed

File: pkg/distributor/distributor.go (hunk 2 above)
Comment (cstyan) [reply pkg/distributor/distributor.go]: Done. I've simplified the function as well so the pointer isn't needed anymore.

File: pkg/distributor/distributor.go [hunk 4]
```diff
@@ -249,11 +250,11 @@ func shardByAllLabels(userID string, labels []client.LabelAdapter) (uint32, erro
 	return h, nil
 }
 
-// Remove the replica label from a slice of LabelPairs if it exists.
-func removeReplicaLabel(replica string, labels *[]client.LabelAdapter) {
+// Remove the label labelname from a slice of LabelPairs if it exists.
+func removeLabel(labelName string, labels *[]client.LabelAdapter) {
```
[the remark this answers is not in the document]
Comment (cstyan) [reply pkg/distributor/distributor.go]: I'll have a look at the labelsPairs function in the morning.

**Comment by cstyan** [comment]
@bboreham thanks for the review! I've split out the refactoring into another commit. I also changed the `validateSeries` function to no longer require the error pointer argument in addition to returning an error.

**Comment by bboreham** [comment]
Given that it isn't tied to HA, could the feature be done as a general label-removal mechanism? This would give us a portion of #771.

**Comment by cstyan** [comment]
I moved the cluster label dropping inside the check for `removeReplica`, which is tied to HA. I can move that back out and turn the label removal into a more general non-HA dependent thing, maybe in another PR? if you'd like.

**Comment by bboreham** [comment]
Given that it doesn't cost me anything, yes I would prefer a generic "for tenant X, everywhere you see label Y, drop it".

**Comment by nbellowe** [comment]
@cstyan Any updates with this PR? Its much appreciated, we're waiting on this to enable HA writing to cortex.

File: docs/arguments.md [hunk 5]
```diff
@@ -144,11 +144,13 @@ prefix these flags with `distributor.ha-tracker.`
 
 ### HA Tracker
 
-HA tracking has two of it's own flags:
+HA tracking has three of its own flags:
 - `distributor.ha-tracker.cluster`
    Prometheus label to look for in samples to identify a Prometheus HA cluster. (default "cluster")
 - `distributor.ha-tracker.replica`
    Prometheus label to look for in samples to identify a Prometheus HA replica. (default "`__replica__`")
+- `distributor.ha-tracker.drop-cluster-label`
```
Comment (bboreham) [inline docs/arguments.md]: does this name need to be updated?

File: pkg/util/validation/limits.go [hunk 6]
```diff
@@ -48,8 +51,9 @@ func (l *Limits) RegisterFlags(f *flag.FlagSet) {
 	f.Float64Var(&l.IngestionRate, "distributor.ingestion-rate-limit", 25000, "Per-user ingestion rate limit in samples per second.")
 	f.IntVar(&l.IngestionBurstSize, "distributor.ingestion-burst-size", 50000, "Per-user allowed ingestion burst size (in number of samples). Warning, very high limits will be reset every -distributor.limiter-reload-period.")
 	f.BoolVar(&l.AcceptHASamples, "distributor.ha-tracker.enable-for-all-users", false, "Flag to enable, for all users, handling of samples with external labels identifying replicas in an HA Prometheus setup.")
-	f.StringVar(&l.HAReplicaLabel, "distributor.ha-tracker.replica", "__replica__", "Prometheus label to look for in samples to identify a Prometheus HA replica.")
 	f.StringVar(&l.HAClusterLabel, "distributor.ha-tracker.cluster", "cluster", "Prometheus label to look for in samples to identify a Prometheus HA cluster.")
+	f.StringVar(&l.HAReplicaLabel, "distributor.ha-tracker.replica", "__replica__", "Prometheus label to look for in samples to identify a Prometheus HA replica.")
```
Comment (bboreham) [inline pkg/util/validation/limits.go]: It's not obvious to me why this moved.  No big deal.

File: docs/arguments.md (hunk 5 above)
Comment (cstyan) [reply docs/arguments.md]: Yes, thanks for catching that.

File: pkg/util/validation/limits.go (hunk 6 above)
Comment (cstyan) [reply pkg/util/validation/limits.go]: Sorry for the noise. I just rearranged the calls to `flags.VarType` here for the cluster and replica labels to match the order those values are specified in the Limits struct.

File: pkg/distributor/distributor.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (cstyan) [reply pkg/distributor/distributor.go]: `labelPairs` is nested within the `ingester` package. If we're going to merge the two functions should that exist in a new package under `util`?

And the reason for preserving order is only that I've generally seen that be the case within Prometheus itself. For example within remote write, when processing a sample through external labelse and relabel rules, we preserve the sorted order of the labels.

File: pkg/distributor/distributor.go [hunk 7]
```diff
@@ -280,6 +281,48 @@ func (d *Distributor) checkSample(ctx context.Context, userID, cluster, replica
 	return true, nil
 }
 
+// Validates a single series from a write request. Will remove HA labels if necessary.
+// Takes a pointer for a partial error so that we can get partial errors, errors during validation
```
Comment (gouthamve) [inline pkg/distributor/distributor.go]: Can you update the comment to remove the reference to the pointer?

File: pkg/util/validation/limits.go [hunk 8]
```diff
@@ -142,14 +146,19 @@ func (o *Overrides) AcceptHASamples(userID string) bool {
...
+
 // HAReplicaLabel returns the replica label to look for when deciding whether to accept a sample from a Prometheus HA replica.
 func (o *Overrides) HAReplicaLabel(userID string) string {
 	return o.overridesManager.GetLimits(userID).(*Limits).HAReplicaLabel
 }
 
-// HAClusterLabel returns the cluster label to look for when deciding whether to accept a sample from a Prometheus HA replica.
-func (o *Overrides) HAClusterLabel(userID string) string {
-	return o.overridesManager.GetLimits(userID).(*Limits).HAClusterLabel
+// DropLabels returns whether the cluster label should be dropped when ingesting HA samples for the user.
```
Comment (gouthamve) [inline pkg/util/validation/limits.go]: Returns the list of labels to be dropped?

File: pkg/distributor/distributor.go [hunk 9]
```diff
@@ -280,6 +281,48 @@ func (d *Distributor) checkSample(ctx context.Context, userID, cluster, replica
...
+// of a single sample, back from the function without adding an additional return param.
+// Returns a token for the series if it is valid, the validated series with it's labels/samples, and any fatal error.
+func (d *Distributor) validateSeries(key uint32, ts ingester_client.PreallocTimeseries, userID string, removeReplica bool) (client.PreallocTimeseries, error) {
+	// If we found both the cluster and replica labels, we only want to include the cluster label when
+	// storing series in Cortex. If we kept the replica label we would end up with another series for the same
+	// series we're trying to dedupe when HA tracking moves over to a different replica.
+	if removeReplica {
+		removeLabel(d.limits.HAReplicaLabel(userID), &ts.Labels)
+	}
+	for _, s := range d.limits.DropLabels(userID) {
```
Comment (gouthamve) [inline pkg/distributor/distributor.go]: `s` is a weird variable name in this instance, maybe call it `label`/`labelName`?

File: pkg/distributor/distributor.go [hunk 10]
```diff
@@ -280,6 +281,48 @@ func (d *Distributor) checkSample(ctx context.Context, userID, cluster, replica
...
+		removeLabel(s, &ts.Labels)
+	}
+	key, err := d.tokenForLabels(userID, ts.Labels)
```
Comment (gouthamve) [inline pkg/distributor/distributor.go]: Also check and handle if the series has no labels anymore after the removing of labels.

**Comment by nbellowe** [comment]
Thank you @cstyan!

**Pull request merged**